first_task.scala

```diff
@@ -0,0 +1,260 @@
+println(countMeanAndMedianOfAnswerScores(answersDs,containsLink(_)))
+println(countMeanAndMedianOfAnswerScores(answersDs, !containsLink(_)))
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Code length (lines) distribution (visualize)
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Распределение по графику похоже на геометрическое, в котором p --- вероятность, того, что user закончит вводить код(включая "не начинание ввода"). Проверка гипотезы будет добавлена 
+
+// COMMAND ----------
+
+val codeLengthProvider: (String => Int) = Jsoup.parse(_).select("code").html.length
+val codeLenDf = answersDs.select(col("body"))
+  .unionAll(questionsDs.select(col("body")))
+  .map(r => codeLengthProvider(r(0).toString))
+
+display(codeLenDf)
+
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Crrelation between the code length and answer score (visualize)
+
+// COMMAND ----------
+
+val codeLenScoreDf = answersDs
+    .map(a => (a.score, Jsoup.parse(a.body).select("code").html.length))
+    .select(col("_1").alias("score"), col("_2").alias("length"))
+
+display(codeLenScoreDf)
+      
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Statistics on the major topics (top 100 tags de-duped), such as django, flask, numpy, pandas etc.
+
+// COMMAND ----------
+
+val majorTopics = tagsDs
+      .groupBy("tag")
+      .agg(count("id").alias("amount"))
+      .orderBy(desc("amount"))
+      .limit(100)
+
+majorTopics.show()
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Experts (users with the best score) per major topics 
+
+// COMMAND ----------
+
+val totalTopicScoreDf = majorTopics
+      .join(tagsDs, Seq("tag"), "left_outer")
+      .join(questionsDs, Seq("id"), "inner")
+      .groupBy("tag", "ownerUserId")
+      .agg(sum("score").alias("total_score"))
+
+totalTopicScoreDf
+  .groupBy("tag")
+  .agg(max("total_score").alias("total_score"))
+  .join(totalTopicScoreDf, Seq("tag", "total_score"), "inner")
+  .orderBy(desc("total_score"))
+  .limit(100)
+  .show()
+
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC What are the most popular (citated) resources referenced in the answers (top 100)
+
+// COMMAND ----------
+
+answersDs
+   .filter(a => !Jsoup.parse(a.body).select("a[href]").isEmpty)
+   .map(a => {
+   try {
+     val url = Jsoup.parse(a.body).select("a[href]").html
+     new URI(url).getHost
+     }  catch {
+       case ex: URISyntaxException => null
+     }
+   })
+   .na
+   .drop("all")
+   .select(col("value").alias("host"))
+   .groupBy("host")
+   .agg(count("*").alias("amount_of_references"))
+   .orderBy(desc("amount_of_references"))
+   .limit(100)
+   .show()
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Seasonality for both questions and answers (yearly)
+
+// COMMAND ----------
+
+val selectCols = Seq($"id", year($"creationDate").alias("year"))
+questionsDs
+  .select(selectCols: _*)
+  .union(answersDs.select(selectCols: _*))
+  .groupBy("year")
+  .agg(count("*").alias("amount_of_questions_answers"))
```

Comment:
this is not quite right in terms of what I mean by seasonality. 
https://en.wikipedia.org/wiki/Seasonality
We analyze all the years per smaller interval, such as day/week or month to understand the trends.
For example, during the summer-time or Christmas time there are less questions posted.
optional: you might be also interested in daily trends (activity peaks etc)